adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/AmqpContext.java

```diff
@@ -63,21 +63,21 @@ ProtonDelivery delivery() {
     }
 
     /**
-     * Gets the tenant identifier of this context's resource.
+     * Gets the tenant identifier for this context.
      *
      * @return The tenant identifier.
      */
     String getTenantId() {
-        return resource.getTenantId();
+        return isDeviceAuthenticated() ? authenticatedDevice.getTenantId() : resource.getTenantId();
```

Comment:
While this will probably yield the intended result, I would rather try to make sure in the receiver open handler that the receiver link's target address actually contains a tenant **and** device if the device is not authenticated. For an authenticated device, IMHO it should also check, if the target address contains either no tenant and device ID at all or a tenant ID that matches the authenticated device (this is relevant for gateways connecting to the adapter).
The receiver open handler seems to suggest, that anonymous relay mode is not supported. However, I think that for gateway mode, it would actually be very helpful to support this because otherwise a gateway would need to create separate links for each device it wants to publish data for.

Reply:
> However, I think that for gateway mode, it would actually be very helpful to support this because otherwise a gateway would need to create separate links for each device it wants to publish data for.

But shouldn't the gateway not indicate at least the `endpoint` that it wants to publish data to (either telemetry or event)?

Reply:
what would be the benefit?

Reply:
to identify if the gateway wants to publish an event or a telemetry message. or maybe am missing something?

Reply:
the endpoint, tenant and device ID will be contained in the message's *to* property in that case, i.e. you can determine with each message being received, whether it is a telemetry message or an event ...

Reply:
ok, i see now.. thanks! 

i think i will have to find a different CLI client for testing with gateways, since `qpid-send` mandates that the link address be provided via the `-a` option